codespan-reporting/examples/custom_files.rs

```diff
@@ -146,8 +127,20 @@ mod files {
             line_index: usize,
         ) -> Result<Range<usize>, files::Error> {
             let file = self.get(file_id)?;
-            let line_start = file.line_start(line_index)?;
-            let next_line_start = file.line_start(line_index + 1)?;
+            let line_start =
+                file.line_starts
+                    .get(line_index)
+                    .copied()
+                    .ok_or(files::Error::LineTooLarge {
+                        given: line_index,
+                        max: file.line_starts.len() - 1,
+                    })?;
+            let next_line_start = file.line_starts.get(line_index + 1).copied().ok_or(
+                files::Error::LineTooLarge {
+                    given: line_index,
+                    max: file.line_starts.len() - 1,
+                },
+            )?;
```

Comment:
Guess I'm a bit concerned at the complexity of this? Like, how easy it is for users to make their own `Files` implementations. We might also need to update the `line_index` implementation here.

Reply:
It is bascially all the error handling that makes this so verbose. But there is no method of the Files trait that this could be moved into since `line_start` has been removed. Also `line_index` fails only if the file is missing and for other errors returns the byte index at the end of the file and that should probably not be changed.

Maybe we should also think about the opportunities this gives to implementers which goes in the direction of enabling caching and such? (*It's not a bug, it's a feature...*)